rust/arrow/src/compute/cast_kernels.rs

```diff
@@ -0,0 +1,543 @@
+        (UInt8, UInt16) => cast_numeric_arrays!(array, UInt8Type, UInt16Type),
+        (UInt8, UInt32) => cast_numeric_arrays!(array, UInt8Type, UInt32Type),
+        (UInt8, UInt64) => cast_numeric_arrays!(array, UInt8Type, UInt64Type),
+        (UInt8, Int8) => cast_numeric_arrays!(array, UInt8Type, Int8Type),
+        (UInt8, Int16) => cast_numeric_arrays!(array, UInt8Type, Int16Type),
+        (UInt8, Int32) => cast_numeric_arrays!(array, UInt8Type, Int32Type),
+        (UInt8, Int64) => cast_numeric_arrays!(array, UInt8Type, Int64Type),
+        (UInt8, Float32) => cast_numeric_arrays!(array, UInt8Type, Float32Type),
+        (UInt8, Float64) => cast_numeric_arrays!(array, UInt8Type, Float64Type),
+
+        (UInt16, UInt8) => cast_numeric_arrays!(array, UInt16Type, UInt8Type),
+        (UInt16, UInt32) => cast_numeric_arrays!(array, UInt16Type, UInt32Type),
+        (UInt16, UInt64) => cast_numeric_arrays!(array, UInt16Type, UInt64Type),
+        (UInt16, Int8) => cast_numeric_arrays!(array, UInt16Type, Int8Type),
+        (UInt16, Int16) => cast_numeric_arrays!(array, UInt16Type, Int16Type),
+        (UInt16, Int32) => cast_numeric_arrays!(array, UInt16Type, Int32Type),
+        (UInt16, Int64) => cast_numeric_arrays!(array, UInt16Type, Int64Type),
+        (UInt16, Float32) => cast_numeric_arrays!(array, UInt16Type, Float32Type),
+        (UInt16, Float64) => cast_numeric_arrays!(array, UInt16Type, Float64Type),
+
+        (UInt32, UInt8) => cast_numeric_arrays!(array, UInt32Type, UInt8Type),
+        (UInt32, UInt16) => cast_numeric_arrays!(array, UInt32Type, UInt16Type),
+        (UInt32, UInt64) => cast_numeric_arrays!(array, UInt32Type, UInt64Type),
+        (UInt32, Int8) => cast_numeric_arrays!(array, UInt32Type, Int8Type),
+        (UInt32, Int16) => cast_numeric_arrays!(array, UInt32Type, Int16Type),
+        (UInt32, Int32) => cast_numeric_arrays!(array, UInt32Type, Int32Type),
+        (UInt32, Int64) => cast_numeric_arrays!(array, UInt32Type, Int64Type),
+        (UInt32, Float32) => cast_numeric_arrays!(array, UInt32Type, Float32Type),
+        (UInt32, Float64) => cast_numeric_arrays!(array, UInt32Type, Float64Type),
+
+        (UInt64, UInt8) => cast_numeric_arrays!(array, UInt64Type, UInt8Type),
+        (UInt64, UInt16) => cast_numeric_arrays!(array, UInt64Type, UInt16Type),
+        (UInt64, UInt32) => cast_numeric_arrays!(array, UInt64Type, UInt32Type),
+        (UInt64, Int8) => cast_numeric_arrays!(array, UInt64Type, Int8Type),
+        (UInt64, Int16) => cast_numeric_arrays!(array, UInt64Type, Int16Type),
+        (UInt64, Int32) => cast_numeric_arrays!(array, UInt64Type, Int32Type),
+        (UInt64, Int64) => cast_numeric_arrays!(array, UInt64Type, Int64Type),
+        (UInt64, Float32) => cast_numeric_arrays!(array, UInt64Type, Float32Type),
+        (UInt64, Float64) => cast_numeric_arrays!(array, UInt64Type, Float64Type),
+
+        (Int8, UInt8) => cast_numeric_arrays!(array, Int8Type, UInt8Type),
```

Comment:
For these casts where the physical storage format is the same, can we just change the data type associated with `ArrayData` such as [how it is done in cpp ](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array.h#L65).

Reply:
Would this be for going from `UInt8` to `Int{8|16|32|64}` and not the inverse like `Int{x}` -> `UInt{x}`? I've just done a quick test:

```rust
    fn test_associated_data() {
        let a = Int32Array::from(vec![-5, 6]);
        let b: PrimitiveArray<UInt32Type> = PrimitiveArray::from(a.data());
        let c: PrimitiveArray<Int32Type> = PrimitiveArray::from(b.data());

        // I expect -5 to be invalid for UInt8, not sure how I'd change the bitmask 
        assert_eq!(false, b.is_valid(0));
        assert_eq!(6, b.value(1));

        // this is also not the case as I still get -5
        assert_eq!(false, c.is_null(0));
        assert_eq!(6, c.value(1));
    }
```

I expect a negative Int cast to UInt to have `is_valid(x) == false`, but because I haven't changed the bitmask, that's not the case.

I'll have a look at how I could do that if I need to cast from signed to unsigned. I also wouldn't mind getting help with those (and other more efficient) casts in follow-up JIRAs.